### PR TITLE
Added Heirarchical prior.

### PR DESCRIPTION
**Pull request by meetdoshi90**
Added path smoothing to prevent overfitting for deeper trees whilst improving accuracy.
Added mse loss for regression on weighted layers.
References: 
1. [Path smoothing](https://ojs.uma.ac.id/index.php/jite/article/view/3856)
2. [IHT Gradient descent](https://arxiv.org/pdf/1910.13389.pdf)

### CODE REVIEW

**Comment by adityakusupati** [comment]
hi @meetdoshi90,

This is very useful and seems correct. Thanks for the contribution. However, I think the TF bonsai version is obsolete due to its old dependencies. I think more people use the pytorch version more often. 

@harsha-simhadri you might want to take a look at this when you get a chance. 

Aditya

**Comment by meetdoshi90** [comment]
Hey @adityakusupati 
I will implement the same on pytorch. Thanks for the advice.
Meet.
